models/siegert_neuron.cpp

```diff
@@ -0,0 +1,479 @@
+
+  if ( tau_syn_ < 0 )
+    throw BadProperty( "Membrane time constant must not be negative." );
+}
+
+void
+nest::siegert_neuron::State_::get( DictionaryDatum& d ) const
+{
+  def< double >( d, names::rate, r_ ); // Rate
+}
+
+void
+nest::siegert_neuron::State_::set( const DictionaryDatum& d )
+{
+  updateValue< double >( d, names::rate, r_ ); // Rate
+}
+
+nest::siegert_neuron::Buffers_::Buffers_( siegert_neuron& n )
+  : logger_( n )
+{
+}
+
+nest::siegert_neuron::Buffers_::Buffers_( const Buffers_&, siegert_neuron& n )
+  : logger_( n )
+{
+}
+
+/* ----------------------------------------------------------------
+ * Default and copy constructor for node
+ * ---------------------------------------------------------------- */
+
+nest::siegert_neuron::siegert_neuron()
+  : Archiving_Node()
+  , P_()
+  , S_()
+  , B_( *this )
+{
+  recordablesMap_.create();
+}
+
+nest::siegert_neuron::siegert_neuron( const siegert_neuron& n )
+  : Archiving_Node( n )
+  , P_( n.P_ )
+  , S_( n.S_ )
+  , B_( n.B_, *this )
+{
+}
+
+/* ----------------------------------------------------------------
+ * Node initialization functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::siegert_neuron::init_state_( const Node& proto )
+{
+  const siegert_neuron& pr = downcast< siegert_neuron >( proto );
+  S_ = pr.S_;
+}
+
+void
+nest::siegert_neuron::init_buffers_()
+{
+  // resize buffers
+  const size_t quantity = kernel().connection_manager.get_min_delay();
+  B_.drift_input_.resize( quantity, 0.0 );
+  B_.diffusion_input_.resize( quantity, 0.0 );
+  B_.last_y_values.resize( quantity, 0.0 );
+
+  B_.logger_.reset(); // includes resize
+  Archiving_Node::clear_history();
+}
+
+void
+nest::siegert_neuron::calibrate()
+{
+  B_.logger_
+    .init(); // ensures initialization in case mm connected after Simulate
+
+  const double h = Time::get_resolution().get_ms();
+
+  // propagators
+  V_.P1_ = std::exp( -h / P_.tau_ );
+  V_.P2_ = -numerics::expm1( -h / P_.tau_ );
+}
+
+/* ----------------------------------------------------------------
+ * Update and event handling functions
+ */
+
+bool
+nest::siegert_neuron::update_( Time const& origin,
+  const long from,
+  const long to,
+  const bool wfr_update )
+{
+  assert(
+    to >= 0 && ( delay ) from < kernel().connection_manager.get_min_delay() );
+  assert( from < to );
+
+  bool done = true;
+  const size_t quantity = kernel().connection_manager.get_min_delay();
+  const double wfr_tol = kernel().simulation_manager.get_wfr_tol();
+
+  // allocate memory to store rates to be sent by rate events
+  std::vector< double > new_rates( quantity, 0.0 );
+
+  for ( long lag = from; lag < to; ++lag )
+  {
+    // register rate in buffer
+    new_rates[ lag ] = S_.r_;
+
+    // propagate rate to new time step (exponential integration)
+    double drive = siegert( P_.tau_m_ * 1e-3,
+      P_.tau_syn_ * 1e-3,
+      P_.t_ref_ * 1e-3,
```

Comment:
Good point. The output of the function siegert is a rate in units of Hz. The NEST convention of time constants in ms would therefore give kHz for rates. We improved the conversion and shifted it to the return statement of the siegert functions.